File: deno/day08.ts
import { DaySolution, parseData } from './_tools.ts';
import { parseLines } from './_helpers.ts';

class Position {
  constructor(
    public x: number,
    public y: number,
  ) {}

  toString() {
    return `${this.x.toString().padStart(2, ' ')},${this.y.toString().padStart(2, ' ')}`;
  }
}

class Data {
  all = new Set<Position>();
  indexed = new Map<string, Set<Position>>();

  width = -1;
  height = -1;

  constructor(
    grid = [] as string[][],
  ) {
    this.width = grid[0].length;
    this.height = grid.length;
    for (let y = 0; y < grid.length; y++) {
      for (let x = 0; x < grid[0].length; x++) {
        const value = grid[y][x];
        if (value !== '.') {
          this.add(value, new Position(x, y));
        }
      }
    }
  }

  add(value: string, position: Position) {
    this.all.add(position);
    if (!this.indexed.has(value)) {
      this.indexed.set(value, new Set());
    }
    this.indexed.get(value)!.add(position);
  }

  isEmpty({ x, y }: Position): boolean {
    return !this.all.has(new Position(x, y));
  }

  inGrid({ x, y }: Position): boolean {
    return x >= 0 && x < this.width && y >= 0 && y < this.height;
  }

  validPosition({ x, y }: Position): boolean {
    return this.inGrid({ x, y }) && this.isEmpty(new Position(x, y));
  }

  calculateAntinodesPart1(p1: Position, p2: Position): Position[] {
    const dx = p2.x - p1.x;
    const dy = p2.y - p1.y;

    if (dx === 0 || dy === 0) {
      return [];
    }

    const antinode1 = new Position(p2.x + dx, p2.y + dy);
    const antinode2 = new Position(p1.x - dx, p1.y - dy);
    return [antinode1, antinode2].filter((p) => this.validPosition(p));
  }

  calculateAntinodesPart2(p1: Position, p2: Position): Position[] {
    const dx = p2.x - p1.x;
    const dy = p2.y - p1.y;

    if (dx === 0 || dy === 0) {
      return [];
    }

    const res = [] as Position[];

    let antinode1 = p2;
    while (this.inGrid(antinode1)) {
      res.push(antinode1);
      antinode1 = new Position(antinode1.x + dx, antinode1.y + dy);
    }

    let antinode2 = p1;
    while (this.inGrid(antinode2)) {
      res.push(antinode2);
      antinode2 = new Position(antinode2.x - dx, antinode2.y - dy);
    }

    return res;
  }
}

function solve(data: Data, calculator: (p1: Position, p2: Position) => Position[]): number {
  return [...data.indexed.entries()]
    .map(([, antennas]) => {
      let res = [] as Position[];
      for (const p1 of antennas) {
        for (const p2 of antennas) {
          res = [...res, ...calculator(p1, p2)];
        }
      }
      return res;
    })
    .flat()
    .reduce((antinodes, antinode) => {
      antinodes.add(antinode.toString());
      return antinodes;
    }, new Set<string>())
    .size;
}

export class Day08 implements DaySolution<Data, number> {
  day = 8;

  parseData(dataLabel: 'Example' | 'Input'): Data {
    const grid = parseLines<string[]>(parseData(this.day, dataLabel), (line) => {
      return line.split('');
    });
    return new Data(grid);
  }

  part1(data: Data): number {
    return solve(data, (p1, p2) => data.calculateAntinodesPart1(p1, p2));
  }

  part2(data: Data): number {
    return solve(data, (p1, p2) => data.calculateAntinodesPart2(p1, p2));
  }
}

File: rust/src/day08.rs
use crate::tools::read_raw_data;
use crate::DaySolution;
use std::collections::{HashMap, HashSet};
use std::ops::{Add, Sub};

#[derive(Clone, Copy, Hash, Eq, PartialEq)]
struct Position {
    x: i32,
    y: i32,
}

impl Position {
    fn new(x: i32, y: i32) -> Self {
        Self { x, y }
    }
}

impl Add for Position {
    type Output = Position;

    fn add(self, rhs: Self) -> Self::Output {
        Position::new(self.x + rhs.x, self.y + rhs.y)
    }
}

impl Sub for Position {
    type Output = Position;

    fn sub(self, rhs: Self) -> Self::Output {
        Position::new(self.x - rhs.x, self.y - rhs.y)
    }
}

pub struct Data {
    width: i32,
    height: i32,
    all: HashSet<Position>,
    indexed: HashMap<char, HashSet<Position>>,
}

impl Data {
    fn new(width: i32, height: i32) -> Self {
        Self {
            width,
            height,
            all: HashSet::new(),
            indexed: HashMap::new(),
        }
    }

    fn add_antenna(&mut self, c: char, x: i32, y: i32) {
        let pos = Position::new(x, y);
        self.all.insert(pos);
        self.indexed
            .entry(c)
            .or_insert(HashSet::new())
            .insert(pos);
    }

    fn in_grid(&self, pos: &Position) -> bool {
        pos.x >= 0 && pos.x < self.width && pos.y >= 0 && pos.y < self.height
    }

    fn calculate_antinodes(&self, antenna1: Position, antenna2: Position, with_resonance: bool) -> Vec<Position> {
        let diff = antenna2 - antenna1;
        if diff.x == 0 && diff.y == 0 {
            return Vec::new();
        }

        let mut antinodes = Vec::new();
        let mut antinode: Position;

        if !with_resonance {
            antinode = antenna2 + diff;
            if self.in_grid(&antinode) {
                antinodes.push(antinode);
            }
            antinode = antenna1 - diff;
            if self.in_grid(&antinode) {
                antinodes.push(antinode);
            }
        } else {
            antinode = antenna2.clone();
            while self.in_grid(&antinode) {
                antinodes.push(antinode);
                antinode = antinode + diff;
            }
            antinode = antenna1.clone();
            while self.in_grid(&antinode) {
                antinodes.push(antinode);
                antinode = antinode - diff;
            }
        }

        antinodes
    }

    fn calculate_all_antinodes(&self, with_resonance: bool) -> HashSet<Position> {
        self.indexed.values()
            .flat_map(|antennas| {
                antennas.iter().flat_map(|antenna1| {
                    antennas.iter().flat_map(move |antenna2| {
                        self.calculate_antinodes(*antenna1, *antenna2, with_resonance)
                    })
                })
            })
            .collect::<HashSet<Position>>()
    }
}

pub struct Day08;

impl DaySolution<Data, usize> for Day08 {
    fn new() -> Self {
        Self {}
    }

    fn day(&self) -> u16 { 8 }

    fn parse_input(&mut self, example: bool) -> Data {
        let raw_data = read_raw_data(self.day(), example);
        let raw_data = raw_data.lines().collect::<Vec<&str>>();

        let mut data = Data::new(raw_data[0].len() as i32, raw_data.len() as i32);

        raw_data.iter().enumerate().for_each(|(y, line)| {
            line.chars().enumerate().for_each(|(x, c)| {
                if c != '.' {
                    data.add_antenna(c, x as i32, y as i32);
                }
            });
        });

        data
    }

    fn part1(&mut self, input: &Data) -> usize {
        input.calculate_all_antinodes(false).len()
    }

    fn part2(&mut self, input: &Data) -> usize {
        input.calculate_all_antinodes(true).len()
    }
}
